OpenSim/Common/Component.cpp

```diff
@@ -865,11 +880,29 @@ SimTK::Vector Component::
     getStateVariableValues(const SimTK::State& state) const
 {
     int nsv = getNumStateVariables();
-    Array<std::string> names = getStateVariableNames();
+    // Consider the list of all StateVariables to be valid if all of 
+    // the following are true:
+    // 1. Component is up-to-date with its Properties
+    // 2. a System has been associated with the list of StateVariables
+    // 3. The list of all StateVariables is correctly sized (initialized)
+    // 4. The System associated with the StateVariables is the current System
+    bool valid = isObjectUpToDateWithProperties() &&   // 1.
+                _statesAssociatedSystem.empty()  &&     // 2.
+                _allStateVariables.size() == nsv &&     // 3.
+                getSystem().isSameSystem(_statesAssociatedSystem.getRef());// 4.
+    // if the StateVariables are invalid (see above) rebuild the list 
+    if (!valid) {
```

Comment:
I think you are asking for more coverage of the conditions and to do that we can come up with atypical but plausible use cases like:
 1. repeated property changes and performing a brief simulation upon each repeat where answers are known to change
 2. similar to 1. but a body is added (via a joint) so that the number of states is changing
 3. purposed wiping out the system (1. and 2. should have this as a side-effect)

Let me know if you have additional cases you would like me to try.
Should these be baked into testComponents or testComponentInterface or make a new test?